pkg/stacker/referrer.go

```diff
@@ -0,0 +1,442 @@
+
+	subject := distspecManifestURL(registry, strings.Split(repo, ":")[0], subjectTag, skipTLS)
+
+	// check subject exists
+	res, err := clientRequest(http.MethodHead, subject, username, password, nil, nil)
+	if err != nil {
+		log.Errorf("unable to check subject:%s, err:%s", subject, err)
+		return err
+	}
+	if res == nil || res.StatusCode != http.StatusOK {
+		log.Errorf("subject:%s doesn't exist, ignoring and proceeding", subject)
+	}
+
+	slen := res.ContentLength
+	smtype := res.Header.Get("Content-Type")
+	sdgst, err := godigest.Parse(res.Header.Get("Docker-Content-Digest"))
+	if slen < 0 || smtype == "" || sdgst == "" || err != nil {
+		log.Errorf("unable to get descriptor details for subject:%s", subject)
+		return errors.Errorf("unable to get descriptor details for subject:%s", subject)
+	}
+
+	// upload the artifact
+	finfo, err := os.Lstat(path)
+	if err != nil {
+		log.Errorf("unable to stat file:%s, err:%s", path, err)
+		return err
+	}
+
+	dgst, err := fileDigest(path)
+	if err != nil {
+		log.Errorf("unable get digest for file:%s, err:%s", path, err)
+		return err
+	}
+
+	fh, err := os.Open(path)
+	if err != nil {
+		log.Errorf("unable to open file:%s, err:%s", path, err)
+		return err
+	}
+	defer fh.Close()
+
+	if err := uploadBlob(registry, repo, path, username, password, fh, finfo.Size(), dgst, skipTLS); err != nil {
+		log.Errorf("unable to upload file:%s, err:%s", path, err)
+		return err
+	}
+
+	// check and upload emptyJSON blob
+	erdr := bytes.NewReader(ispec.DescriptorEmptyJSON.Data)
+	edgst := ispec.DescriptorEmptyJSON.Digest
+
+	if err := uploadBlob(registry, repo, path, username, password, erdr, ispec.DescriptorEmptyJSON.Size, &edgst, skipTLS); err != nil {
+		log.Errorf("unable to upload file:%s, err:%s", path, err)
+		return err
+	}
+
+	// upload the reference manifest
+	manifest := ispec.Manifest{
+		Versioned: specs.Versioned{
+			SchemaVersion: 2,
+		},
+		MediaType:    ispec.MediaTypeImageManifest,
+		ArtifactType: mtype,
+		Config:       ispec.DescriptorEmptyJSON,
+		Subject: &ispec.Descriptor{
+			MediaType: ispec.MediaTypeImageManifest,
+			Size:      slen,
+			Digest:    sdgst,
+		},
+		Layers: []ispec.Descriptor{
+			ispec.Descriptor{
+				MediaType: mtype,
+				Size:      finfo.Size(),
+				Digest:    *dgst,
+			},
+		},
+	}
+
+	//content, err := json.MarshalIndent(&manifest, "", "\t")
+	content, err := json.Marshal(&manifest)
+	if err != nil {
+		log.Errorf("unable to marshal image manifest, err:%s", err)
+		return err
+	}
+
+	// artifact manifest
+	mdgst := godigest.FromBytes(content)
+	regUrl := distspecManifestURL(registry, strings.Split(repo, ":")[0], mdgst.String(), skipTLS)
+	hdrs := map[string]string{
+		"Content-Type":   ispec.MediaTypeImageManifest,
+		"Content-Length": fmt.Sprintf("%d", len(content)),
+	}
+	res, err = clientRequest(http.MethodPut, regUrl, username, password, hdrs, bytes.NewBuffer(content))
+	if err != nil {
+		log.Errorf("unable to check subject:%s, err:%s", subject, err)
```

Comment:
```suggestion
		log.Errorf("unable to PUT subject:%s, err:%s", subject, err)
```
Here we're pushing the subject object, rright?

Reply:
Or maybe we want to say, `failed to put artifact manifest` as the error here?